klippy/extras/output_pin.py

```diff
@@ -72,7 +72,8 @@ def _set_pin(self, print_time, value, cycle_time, is_resend=False):
                     self.resend_timer, self.reactor.NEVER)
             else:
                 self.reactor.update_timer(
-                    self.resend_timer, self.reactor.monotonic())
+                    self.resend_timer, self.reactor.monotonic()
```

Comment:
I am not sure about this one.
self.reactor.monotonic() returns the "current" systime.
But the `print_time` may be somewhere in the future (around the lookahead time).
Is it here better to instead get the estimated systime at the print_time @KevinOConnor?
e.g.:
```python
eventtime = self.mcu_pin.get_mcu()._clocksync.estimate_clock_systime(print_time)
self.reactor.update_timer(
    self.resend_timer, eventtime + (0.8 * self.host_ack_timeout) - PIN_MIN_TIME
)
```

Reply:
Might be simpler to do all the scheduling work in the timer.  For example, just have something like the following in `_set_pin()`:
```python
if self.host_ack_timeout and self.resend_timer is not None:
    self.resend_timer = self.reactor.add_timer(self._resend, self.reactor.NOW)
```

And then do something like the following in `_resend()`:
```python
def _resend(self, eventtime):
    if self.last_value == self.shutdown_value:
        self.reactor.unregister_timer(self.resend_timer)
        self.resend_timer = None
        return self.reactor.NEVER
    systime = self.reactor.monotonic()
    print_time = self.mcu_pin.get_mcu().estimated_print_time(systime)
    time_diff = print_time - (self.last_print_time + .8 * self.host_ack_timeout - PIN_MIN_TIME)
    if time_diff > 0.:
        # Reschedule for resend time
        return systime + time_diff
    self._set_pin(print_time + PIN_MIN_TIME, self.last_value, self.last_cycle_time, True)
    return systime + (0.8 * self.host_ack_timeout) - PIN_MIN_TIME
```

Reply:
I can try this code tomorrow.
My first thought is, that when a lot of pin updates come (With my "slow" laser setup at ~500 updates/s), will there be 500 spawned timers concurrently rescheduling themselves?

Currently, the timer keeps on being rescheduled upon these updates, without being fired in a dense update phase.

Reply:
This demo code only adds the timer if no timer is already running.  It is optimized for just the scenario you mentioned - not only will it not add a timer - it doesn't even need to reschedule the timer on each update.  (Instead it "lazily reschedules" when the timer triggers.)

FYI, it's just demo code - I haven't tested it at all.

-Kevin